packages/core/src/getNextSnapshot.ts

```diff
@@ -0,0 +1,56 @@
+import { createEmptyActor } from './actors/index.ts';
+import {
+  AnyActorLogic,
+  AnyActorScope,
+  EventFromLogic,
+  SnapshotFrom
+} from './types.ts';
+
+/**
+ * Determines the next snapshot for the given `actorLogic` based on
+ * the given `snapshot` and `event`.
+ *
+ * If the `snapshot` is `undefined`, the initial snapshot of the
+ * `actorLogic` is used.
+ *
+ * @example
+  ```ts
+  import { getNextSnapshot } from 'xstate';
+  import { trafficLightMachine } from './trafficLightMachine.ts';
+
+  const nextSnapshot = getNextSnapshot(
+    trafficLightMachine, // actor logic
+    undefined, // snapshot (or initial state if undefined)
+    { type: 'TIMER' }); // event object
+
+  console.log(nextSnapshot.value);
+  // => 'yellow'
+
+  const nextSnapshot2 = getNextSnapshot(
+    trafficLightMachine, // actor logic
+    nextSnapshot, // snapshot
+    { type: 'TIMER' }); // event object
+
+  console.log(nextSnapshot2.value);
+  // =>'red'
+  ```
+ */
+export function getNextSnapshot<T extends AnyActorLogic>(
+  actorLogic: T,
+  snapshot: SnapshotFrom<T> | undefined,
+  event: EventFromLogic<T>
+): any {
+  const self = createEmptyActor();
+  const inertActorScope: AnyActorScope = {
+    self,
+    defer: () => {},
+    id: '',
+    logger: () => {},
+    sessionId: '',
+    stopChild: () => {},
+    system: self.system
+  };
+  const resolvedSnapshot =
+    snapshot ?? actorLogic.getInitialSnapshot(inertActorScope, null as any);
+  return actorLogic.transition(resolvedSnapshot, event, inertActorScope);
```

Comment:
One extra problem that I see here is that each time this `getNextSnapshot` is called one ends up with a new `inertActorScope`. While it probably won't create big problems in the vast majority of cases, it feels incoherent that a basic assumption that `self` is the same across the whole "session" doesn't hold up when using this.

Reply:
The preferred solution (IMO) would be to attach `self` to the snapshot.

Reply:
Which again is at odds with snapshot management since you'd force every implementer of the actor logic to attach this to its snapshot ;p The pure/impure "split" in our API is quite challenging 😢 Looking at it from a different angle - the problem that we might notice is that "pure" bits have access to the `actorScope`. If the pure bits were decoupled from it that would simplify some things (but it's also not a change that we can take now anyway...)